frame/dex-router/src/lib.rs

```diff
@@ -0,0 +1,190 @@
+		},
+		DispatchResult, FixedPointOperand,
+	};
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+		type AssetId: FullCodec
+			+ Eq
+			+ PartialEq
+			+ Copy
+			+ MaybeSerializeDeserialize
+			+ Debug
+			+ Default
+			+ TypeInfo
+			+ Ord;
+		type Balance: Default
+			+ Parameter
+			+ Codec
+			+ Copy
+			+ Ord
+			+ CheckedAdd
+			+ CheckedSub
+			+ CheckedMul
+			+ AtLeast32BitUnsigned
+			+ From<u64> // at least 64 bit
+			+ Zero
+			+ One
+			+ IntegerSquareRoot
+			+ FixedPointOperand
+			+ Into<LiftedFixedBalance>
+			+ Into<u128>; // cannot do From<u128>, until LiftedFixedBalance integer part is larger than 128
+			  // bit
+		/// The maximum hops in the route.
+		type MaxHopsInRoute: Get<u32> + TypeInfo;
+		type PoolId: FullCodec
+			+ Default
+			+ TypeInfo
+			+ Eq
+			+ PartialEq
+			+ Ord
+			+ Copy
+			+ Debug
+			+ CheckedAdd
+			+ Zero
+			+ One;
+		type PoolTokenIndex: Copy + Debug + Eq + Into<u32>;
+		type StableSwapDex: CurveAmm<
+			AssetId = Self::AssetId,
+			Balance = Self::Balance,
+			AccountId = Self::AccountId,
+			PoolTokenIndex = Self::PoolTokenIndex,
+			PoolId = Self::PoolId,
+		>;
+		type ConstantProductDex: CurveAmm<
+			AssetId = Self::AssetId,
+			Balance = Self::Balance,
+			AccountId = Self::AccountId,
+			PoolTokenIndex = Self::PoolTokenIndex,
+			PoolId = Self::PoolId,
+		>;
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::storage]
+	pub type DexRoutes<T: Config> = StorageDoubleMap<
+		_,
+		Blake2_128Concat,
+		T::AssetId,
+		Blake2_128Concat,
+		T::AssetId,
+		DexRoute<T::PoolId, T::MaxHopsInRoute>,
+		OptionQuery,
+	>;
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Number of hops in route exceeded maximum limit.
+		MaxHopsExceeded,
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		RouteAdded {
+			x_asset_id: T::AssetId,
+			y_asset_id: T::AssetId,
+			route: Vec<DexRouteNode<T::PoolId>>,
+		},
+		RouteDeleted {
+			x_asset_id: T::AssetId,
+			y_asset_id: T::AssetId,
+			route: Vec<DexRouteNode<T::PoolId>>,
+		},
+		RouteUpdated {
+			x_asset_id: T::AssetId,
+			y_asset_id: T::AssetId,
+			route: Vec<DexRouteNode<T::PoolId>>,
+		},
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {}
+
+	impl<T: Config> Pallet<T> {}
+
+	impl<T: Config> DexRouter<T::AssetId, T::PoolId> for Pallet<T> {
+		fn update_route(
+			asset_pair: (T::AssetId, T::AssetId),
+			route: Option<Vec<DexRouteNode<T::PoolId>>>,
+		) -> DispatchResult {
+			let k1 = asset_pair.0;
+			let k2 = asset_pair.1;
+			match route {
+				Some(route_vec) => {
+					let bounded_route =
+						route_vec.clone().try_into().map_err(|_| Error::<T>::MaxHopsExceeded)?;
+					if DexRoutes::<T>::contains_key(k1, k2) {
+						DexRoutes::<T>::insert(k1, k2, DexRoute::Direct(bounded_route));
```

Comment:
let exisitng = ...;
insert();
let event = if exising { update } else { add };
deposit(event)

this males flow more linear and less error prone